lednamebadge.py

```diff
@@ -523,84 +802,267 @@ def header(lengths, speeds, modes, blinks, ants, brightness=100, date=datetime.n
+        Basically it is ready if all necessary libraries and Python modules could be loaded. The method name can be
+        used as a parameter value for write().
+        """
+        auto_order_methods = LedNameBadge._get_auto_order_method_list()
+        return {m.get_name(): (m.get_description(), m.is_ready()) for m in auto_order_methods}
+
+    @staticmethod
+    def get_available_device_ids(method):
+        """Returns all devices available via the given write method as a dict. Each entry has the device id as the key
+        and the device description as the value. The device id can be used as a parameter value for write().
+        """
+        auto_order_methods = LedNameBadge._get_auto_order_method_list()
+        wanted_method = [m for m in auto_order_methods if m.get_name() == method]
+        if wanted_method:
+            return wanted_method[0].get_available_devices()
+        return []
+
+    @staticmethod
+    def _find_write_method(method, device_id):
+        """Here we try to concentrate all special cases, decisions and messages around the manual or automatic
+        selection of write methods and device. This way it is a bit easier to extend or modify the different
+        working run time environments (think of operating system, python version, installed libraries and python
+        modules, ands so on.)"""
+        auto_order_methods = LedNameBadge._get_auto_order_method_list()
+        hidapi = [m for m in auto_order_methods if m.get_name() == 'hidapi'][0]
+        libusb = [m for m in auto_order_methods if m.get_name() == 'libusb'][0]
+
+        if method == 'list':
+            LedNameBadge._print_available_methods(auto_order_methods)
+            sys.exit(0)
+
+        if method not in [m.get_name() for m in auto_order_methods] and method != 'auto':
+            print("Unknown write method '%s'." % (method,))
+            LedNameBadge._print_available_methods(auto_order_methods)
             sys.exit(1)
 
-        if LedNameBadge._have_pyhidapi:
-            dev_info = LedNameBadge.pyhidapi.hid_enumerate(0x0416, 0x5020)
-            # dev = pyhidapi.hid_open(0x0416, 0x5020)
-            if dev_info:
-                dev = LedNameBadge.pyhidapi.hid_open_path(dev_info[0].path)
-                print("using [%s %s] int=%d page=%s via pyHIDAPI" % (
-                    dev_info[0].manufacturer_string, dev_info[0].product_string, dev_info[0].interface_number, dev_info[0].usage_page))
-            else:
-                print("No led tag with vendorID 0x0416 and productID 0x5020 found.")
-                print("Connect the led tag and run this tool as root.")
+        if method == 'auto':
+            if sys.version_info[0] < 3:
+                method = libusb.get_name()
+                print("Preferring method %s over %s with Python 2.x" % (libusb.get_name(), hidapi.get_name()))
+                print("because of https://github.com/jnweiger/led-badge-ls32/issues/9")
+            elif sys.platform.startswith('darwin'):
```

Comment:
Hello @s-light, Jürgen gave me the info, that you might be able to test this new version with MacOs? That would be great!

Reply:
oh sorry - not really..
i have no led-badge myself - and the macos devices are owned by my mother & brother - currently not living near by...

Reply:
Ok, thanks and sorry for bothering...

Reply:
all fine :-)